File: Excelödev/my_py_pkg(4.örnek)/my_py_pkg/pathfollowernode.py
import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose
from geometry_msgs.msg import Twist
import math

class PathFollowerNode(Node):
    def _init_(self):
        super()._init_('path_follower_node')
        self.publisher_ = self.create_publisher(Twist, 'turtle1/cmd_vel', 10)
        self.subscription = self.create_subscription(Pose, 'turtle1/pose', self.pose_callback, 10)
        self.subscription  # To prevent unused variable warning
        self.path = [(2.0, 2.0), (5.0, 5.0), (8.0, 2.0), (5.0, 2.0)]
        self.current_path_index = 0
        self.goal_tolerance = 0.1

    def pose_callback(self, msg):
        x = msg.x
        y = msg.y
        goal_x, goal_y = self.path[self.current_path_index]

        distance_to_goal = math.sqrt((x - goal_x)*2 + (y - goal_y)*2)

        if distance_to_goal <= self.goal_tolerance:
            if self.current_path_index < len(self.path) - 1:
                self.current_path_index += 1
            else:
                self.get_logger().info('Turtlesim hedef yolu tamamladı!')
                self.destroy_node()
                rclpy.shutdown()
        else:
            self.move_to_goal(goal_x, goal_y)

    def move_to_goal(self, goal_x, goal_y):
        twist = Twist()
        twist.linear.x = self.calculate_linear_speed(goal_x, goal_y)
        twist.angular.z = self.calculate_angular_speed(goal_x, goal_y)

        self.publisher_.publish(twist)

    def calculate_linear_speed(self, goal_x, goal_y):
        return 1.0  # İleri hareket hızı

    def calculate_angular_speed(self, goal_x, goal_y):
        x, y, theta = self.get_current_pose()
        desired_theta = math.atan2(goal_y - y, goal_x - x)
        return 1.0 * (desired_theta - theta)  # Dönüş hızı

    def get_current_pose(self):
        # Turtlesim'in mevcut konumunu alır
        # (Bunun için turtlesim için tf2_kdl kütüphanesi gereklidir, ancak siz bu kısmı göstermemişsiniz.)
        return 0.0, 0.0, 0.0  # Bu kısmı geçici olarak boş bıraktım, gerçek konum alınmalıdır

def main(args=None):
    rclpy.init(args=args)
    node = PathFollowerNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Excelödev/my_cpp_pkg/src/pathfollowernode.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <turtlesim/msg/pose.hpp>

class PathFollowerNode : public rclcpp::Node
{
public:
    PathFollowerNode()
        : Node("path_follower_node")
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);
        subscription_ = this->create_subscription<turtlesim::msg::Pose>(
            "turtle1/pose", 10, std::bind(&PathFollowerNode::poseCallback, this, std::placeholders::_1));

        // Belirli bir yol (koordinatlar) tanımlayın
        path_ = {{2.0, 2.0}, {5.0, 5.0}, {8.0, 2.0}, {5.0, 2.0}};
        current_path_index_ = 0;
        goal_tolerance_ = 0.1;
    }

private:
    void poseCallback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        double x = msg->x;
        double y = msg->y;
        double goal_x = path_[current_path_index_].first;
        double goal_y = path_[current_path_index_].second;

        double distance_to_goal = std::sqrt(std::pow(x - goal_x, 2) + std::pow(y - goal_y, 2));

        if (distance_to_goal <= goal_tolerance_)
        {
            if (current_path_index_ < path_.size() - 1)
            {
                current_path_index_++;
            }
            else
            {
                RCLCPP_INFO(this->get_logger(), "Turtlesim hedef yolu tamamladı!");
                rclcpp::shutdown();
            }
        }
        else
        {
            moveTowardGoal(goal_x, goal_y);
        }
    }

    void moveTowardGoal(double goal_x, double goal_y)
    {
        auto twist = std::make_unique<geometry_msgs::msg::Twist>();
        twist->linear.x = calculateLinearSpeed(goal_x, goal_y);
        twist->angular.z = calculateAngularSpeed(goal_x, goal_y);

        publisher_->publish(std::move(twist));
    }

    double calculateLinearSpeed(double goal_x, double goal_y)
    {
        return 1.0; // İleri hareket hızı
    }

    double calculateAngularSpeed(double goal_x, double goal_y)
    {
        auto current_pose = getCurrentPose();
        double desired_theta = std::atan2(goal_y - current_pose[1], goal_x - current_pose[0]);
        return 1.0 * (desired_theta - current_pose[2]); // Dönüş hızı
    }

    std::array<double, 3> getCurrentPose()
    {
        
        return {0.0, 0.0, 0.0}; 
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_;
    std::vector<std::pair<double, double>> path_;
    size_t current_path_index_;
    double goal_tolerance_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PathFollowerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}